php_phongo.c

```diff
@@ -2459,7 +2459,138 @@ static char* php_phongo_manager_make_client_hash(const char* uri_string, zval* o
 	return hash;
 }
 
-static mongoc_client_t* php_phongo_make_mongo_client(const mongoc_uri_t* uri) /* {{{ */
+static bool php_phongo_extract_handshake_data(zval* zvalue, const char* key, char** value, size_t* value_len)
+{
+	if (Z_TYPE_P(zvalue) != IS_STRING) {
+		phongo_throw_exception(PHONGO_ERROR_INVALID_ARGUMENT TSRMLS_CC, "Expected \"%s\" handshake option to be a string, %s given", key, PHONGO_ZVAL_CLASS_OR_TYPE_NAME_P(zvalue));
+		return false;
+	}
+
+	*value     = estrdup(Z_STRVAL_P(zvalue));
+	*value_len = Z_STRLEN_P(zvalue);
+
+	return true;
+}
+
+static char* php_phongo_concat_handshake_data(const char* default_value, const char* custom_value, size_t custom_value_len)
+{
+	char*  ret;
+	size_t ret_len = strlen(default_value) + 1;
+
+	if (custom_value) {
+		ret_len += custom_value_len + 1;
+	}
+
+	ret = malloc(ret_len);
+
+	if (custom_value) {
+		snprintf(ret, ret_len, "%s|%s", default_value, custom_value);
+	} else {
+		snprintf(ret, ret_len, "%s", default_value);
+	}
+
+	return ret;
+}
+
+static void php_phongo_handshake_data_append(const char* name, size_t name_len, const char* version, size_t version_len, const char* platform, size_t platform_len)
+{
+	char* php_version_string;
+	char* driver_name;
+	char* driver_version;
+	char* full_platform;
+
+	php_version_string = malloc(4 + sizeof(PHP_VERSION));
+	sprintf(php_version_string, "PHP %s", PHP_VERSION);
```

Comment:
I realize `sizeof(PHP_VERSION)` includes the null byte and we don't have to worry about overflow, but I think it'd be safer to calculate the length first and then use `snprintf` with the same limit.

Also, can we consider using `ecalloc` and `efree` (or the persistent versions if you prefer)? We don't technically need PHP's memory management, but this will ensure we use consistent allocation functions. As-is, I think this might be the only instance of C's memory allocation functions being used directly. Using PHP's `e` functions would also ensure debug mode catches leaks for us without having to rely on Valgrind for raw leaks.

Reply:
Ah, good point. I've switched to using `ecalloc` and `efree` throughout and am now using `snprintf` and `strlen` to ensure we're not overflowing the previously allocated buffer. Thanks for pointing this out!